hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -553,19 +558,21 @@ public void setPathProperties(final Path path,
    * Orchestrates the copying of blob from given source to a given destination.
    * @param srcPath source path
```

Comment:
For some reason line 301 
`    if (abfsConfiguration.getBlobDirRenameMaxThread() == 0) {
      renameBlobExecutorService = Executors.newFixedThreadPool(
          Runtime.getRuntime()
              .availableProcessors());
    } else {
      renameBlobExecutorService = Executors.newFixedThreadPool(
          abfsConfiguration.getBlobDirRenameMaxThread());
    }`
is not showing as a PR diff to me. Lets go only with the else code with the default value of getBlobDirRenameMaxThread being 5 to match with wasb.

Reply:
What happens to a thread of executorService post thier operation ? Do they remain as blocked threads for this service ? If yes, executor service should not be static and should be created and shutdown for each rename separately.

Reply:
very good comment!
If the queue in ThreadPoolExecutor is empty, the thread will get out of thread-scheduling.
Explanation:
How threadPool is created:
https://github.com/JetBrains/jdk8u_jdk/blob/adfbcc40df1bae8f8a03d249f588fd19f4716c96/src/share/classes/java/util/concurrent/Executors.java#L88-L92, it just adds properties like corePool, maxPool. Threads are not spawned here.
On submit, method `execute` is called which will add workers until corePoolSize is achieved: https://github.com/JetBrains/jdk8u_jdk/blob/00f685db5bd74a44f32992859534f0d8f282040d/src/share/classes/java/util/concurrent/ThreadPoolExecutor.java#L1367.
This will add worker which has a runner and spawn a thread for the worker: https://github.com/JetBrains/jdk8u_jdk/blob/00f685db5bd74a44f32992859534f0d8f282040d/src/share/classes/java/util/concurrent/ThreadPoolExecutor.java#L932-L959
Core of worker: https://github.com/JetBrains/jdk8u_jdk/blob/00f685db5bd74a44f32992859534f0d8f282040d/src/share/classes/java/util/concurrent/ThreadPoolExecutor.java#L623-L625
https://github.com/JetBrains/jdk8u_jdk/blob/00f685db5bd74a44f32992859534f0d8f282040d/src/share/classes/java/util/concurrent/ThreadPoolExecutor.java#L1127
Important point comes here, its a while loop and calls `getTask()` from the while loop:
which makes a call to LinkedBlockingQueue.take() method: https://github.com/JetBrains/jdk8u_jdk/blob/00f685db5bd74a44f32992859534f0d8f282040d/src/share/classes/java/util/concurrent/ThreadPoolExecutor.java#L1072-L1074
Most important implementation here is of take():https://github.com/JetBrains/jdk8u_jdk/blob/09cb295e5810545431c7afcb45bb45081e3887dd/src/share/classes/java/util/concurrent/LinkedBlockingQueue.java#L435-L454
As you can see, if size is 0, the thread gets on condition await.
Once thread is on condition await, it is disabled for thread-scheduling (ref: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/Condition.html#await%E2%80%93)

Reply:
The point that when there is no work is queued up, already spun threads will be idle is understood along with the overhead of polling looking for any work. For a workload pattern where there are no guaranteed on number of folder renames, having such blocked resources is not correct. Shutdown should get called at the end of the rename operation. 
[javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#:~:text=An%20unused%20ExecutorService%20should%20be%20shut%20down%20to%20allow%20reclamation%20of%20its%20resources.)

In an unpredicatable workload patterns were rename isnt an actively called API, we should expect that there will be many a runtime flows where executor service will be unused.

Reply:
Good point! Completely agree on the increased thread count. Have already included this recommendation in last offline review. Now each renameDir will have its own executorService which will be shutdown once whole rename of dir is completed.